compiler/passes/buildDefaultFunctions.cpp

```diff
@@ -76,84 +80,130 @@ static void buildDefaultReadWriteFunctions(AggregateType* type);
 
 static void buildFieldAccessorFunctions(AggregateType* at);
 
+// FunctionExistsMap is a cache of function name -> vector<FnSymbol>
+// We incrementally update it in functionExists as FnSymbol's are
+// only added to gFnSymbols in the duration of this pass
+// We (ugly) manage its lifetime with the pointer sFunctionExistsMap
+// so that the pass can easily cleanup after itself after the entry point
+// buildDefaultFunctions
+// There are other calls into this translation unit that are not part of the
+// pass, so we revert to not using the cache in those circumstances because
+// we can no longer guarantee that gFnSymbols hasn't changed
+class FnSymbolIndex;
+FnSymbolIndex *sFnSymbolIndex = nullptr;
+
+class FnSymbolIndex {
+  std::vector<FnSymbol*> empty_;
+  std::unordered_map<const char *, std::vector<FnSymbol*>> map_;
+  size_t fnSymbolsSize_ = 0;
+
+public:
+  FnSymbolIndex() {sFnSymbolIndex = this;}
+  ~FnSymbolIndex() {sFnSymbolIndex = nullptr;}
+
+  const std::vector<FnSymbol*> &get(const char *astrName) {
+    auto it = map_.find(astrName);
+    if (it == map_.end()) return empty_;
+    return it->second;
+  }
+
+  void update() {
+    size_t n = gFnSymbols.size();
+    if (n > fnSymbolsSize_) {
+      for (size_t i = fnSymbolsSize_; i < n; i++) {
+        FnSymbol *fn = gFnSymbols.v[i];
+        auto it = map_.find(fn->name);
+        if (it == map_.end()) {
+          map_.insert({fn->name, {fn}});
+        } else {
+          it->second.push_back(fn);
+        }
+      }
+    }
+    fnSymbolsSize_ = n;
+  }
+};
+
+
 
 void buildDefaultFunctions() {
+  FnSymbolIndex fsi;
+
   buildChplEntryPoints();
 
   SET_LINENO(rootModule); // todo - remove reset_ast_loc() calls below?
 
-  std::vector<BaseAST*> asts;
-
-  collect_asts(rootModule, asts);
+  // Take a copy of the existing TypeSymbols because the below loop can create
+  // new ones. This maintains compatibility with the previous approach which
+  // used a collect_ast on rootModule
+  std::vector<TypeSymbol *> typeSymbols(gTypeSymbols.begin(), gTypeSymbols.end());
```

Comment:
I don't know that that is accurate - I think this might change behavior?  gTypeSymbols can contain TypeSymbols that are no longer in the AST because they have been replaced or pruned (though I don't think that replacement or removal will have happened at this point in the compiler).  Whereas rootModule will only contain things with a parentExpr or parentSymbol that ultimately leads to the rootModule

Reply:
Yeah I believe this is correct because this is the start of a pass, but I could add an assert that it's in the tree